tests/conftest.py

```diff
@@ -149,11 +149,14 @@ def login_dialog_is_visible():
 
 
 @pytest.fixture(scope="function")
-def functional_test_logged_in_context(functional_test_app_started_context, qtbot):
+def functional_test_logged_in_context(functional_test_app_started_context, qtbot, mocker):
     """
     Returns a tuple containing the gui window and controller of a configured client after logging in
     with our test user account.
     """
+    mocker.patch("securedrop_client.gui.conversation.export.device.Export._run_disk_test")
+    mocker.patch("securedrop_client.gui.conversation.export.device.Export._run_usb_test")
+    mocker.patch("securedrop_client.gui.conversation.export.device.Export._run_disk_export")
```

Comment:
> From there testing the dialog could be done with a "test" device. In practice, with a device which state would be controlled by the test setup, and which (non-GUI) export_service would be mocked.

These three patches in the functional test suite are a less obvious answer to the comment. I think it's worth unpacking why we can't just mock the `export_service` (e.g. by overriding `dialog._device._export_service = mocker.MagicMock(spec=Export)` in the `test_export_dialog` functional  tests).

It has to do with signals (some for thread-safety, some for event-driven UI).

We've got four parts that are relevant to the export feature, three of which are currently explicit.

1. The `export.Dialog` itself, which responsibility is to render UI elements for journalists to interact with. Among other things, the dialog needs to be able to react to changes in the state of the USB drive ("export device"), which can be inserted, unlocked, etc.
2. To achieve that, we've got a `export.Device`. Each dialog holds a "device" object, which responsibility is to keep track of the state of the USB drive.
Because the state of the USB drive can change at any time, the adequate way for the `export.Device` to communicate changes to the `export.Dialog` is through Qt signals. So far, that's what Qt Signals are typically used for. One caveat is that signalling behavior is not easy to mock.
3. The `export.Device` doesn't actually interact with the USB drive. Because the interaction with the USB drive is not GUI concern, it lives in a non-GUI section of the code, which the "device" knows as the `export_service`.
The GUI must be able to react to user input in a given amount of time in order to feel responsive. Because that some operations that the export service performs take longer than the GUI thread can afford to wait, the export service runs is a separate thread. In order to communicate with the `export.Device` in a thread-safe way, the export service used Qt signals (even though it is not GUI code). That is another legitimate use of Qt signals. That comes with the same caveat: the signalling behavior is not easy to mock.
4. Comes in the 4th (currently implicit) actor in play. Besides interfacing with the `export.Device`, the export service interfaces with Qubes OS. That interface is synchronous, and doesn't rely on Qt signals. It is easy to mock, **and it is exactly what we're interested in mocking**. But it is not yet an explicit interface in the way that the other ones are, we can't mock just one object and be done with it. Gladly, the implementation of the export service does encapsulate the Qubes OS interface in a small number of methods, three of which are mocked in this test. (The fact that we haven't mocked other ones... is a flag that tells us that we've got no functional test for printing for example.)

Conclusion: I think these three patches are not the most elegant code, but I think they're justified in the current state of the export code. A desirable next step is to make possible to replace them by a single object (that represents the interface between the export service and Qubes OS - I say Qubes OS, it's the Python representation of the Qubes RPC in practice).

Reply:
Replaced by injecting `export_service`, which is created with the Qubes OS bits disabled in a fixture.